synthtool/languages/python_mono_repo.py

```diff
@@ -98,6 +99,60 @@ def update_url_in_setup_py(package_dir: str):
         f.writelines(new_setup_py)
 
 
+def apply_client_specific_post_processing(
+    client_specific_post_processing_path: str, package_name: str
+) -> None:
+    """Applies client-specific post processing which exists in the Path `client_specific_post_processing_dir`.
+    The client-specific post processing YAML is in the following format:
+    ```
+        description: Verbose description about the need for the workaround.
+        url: URL of the issue in gapic-generator-python tracking eventual removal of the workaround
+        replacements:
+          - replacement:
+            paths: [<List of files where the replacement should occur relative to the monorepo root directory>]
+            before: "The string to search for in the specified paths"
+            after:  "The string to replace in the the specified paths",
+            count: <integer indicating number of replacements that should have occurred across all files after the script is run>
+    ```
+
+    Args:
+        client_specific_post_processing_dir (str): Path to the directory which contains YAML files which will
+            be used to apply client-specific post processing, e.g. 'packages/<package_name>/scripts/client-post-processing'
+            relative to the monorepo root directory.
+        package_name (str): The name of the package where client specific post processing will be applied.
+    """
+
+    if Path(client_specific_post_processing_path).exists():
+        for post_processing_path in Path(
+            client_specific_post_processing_path
+        ).iterdir():
+            with open(post_processing_path, "r") as post_processing_path_file:
+                post_processing_json = yaml.safe_load(post_processing_path_file)
+                all_replacements = post_processing_json["replacements"]
+                # For each workaround related to the specified issue
+                for replacement in all_replacements:
+                    # For each file that needs the workaround applied
+                    for client_library_path in replacement["paths"]:
+                        if package_name in client_library_path:
+                            assert (
+                                synthtool.replace(
+                                    client_library_path,
+                                    replacement["before"],
+                                    replacement["after"],
+                                )
+                                == replacement["count"]
```

Comment:
This still violates idempotency, because `replacement["count"]` is meant to be the sum of all replacements across all the files in `replacement["paths"]`, not just under `package_name`. If we keep those semantics, the assert should not be here, and instead should refer to a (cached?) version of the raw generated files vs the post-processed files.

[Potentially we should change the semantics of `replacement["count"]` to be per `client_library_path`, but looking at the source of `synthtool.replace`, the first argument could be a glob and so the number of replacements could vary by package.]

Reply:
`replacement["paths"]` must only include paths for a single package so that `replacement["count"]` reflects only the replacement for a single package. I fixed the description in https://github.com/googleapis/synthtool/pull/1881/commits/04d32e919280e96cf1a7798e783cdaee44014687. Each package needs to have a separate `-replacement` section in the yaml.

Reply:
Resolving old thread